### PR TITLE
Add assert_show_match/refute_show_match test APIs

### PR DESCRIPTION
**Pull request by glennmatthews**
We have a very common pattern in minitest where we execute some show command over the telnet connection, match it against some regexp pattern, and succeed or fail based on the result. This PR adds a couple of helper methods `assert_show_match` and `refute_show_match` to support this pattern.

```
assert_show_match(command: 'show run all | no-more',
                  pattern: /interface port-channel 1/,
                  msg:     'port-channel is not present but it should be')
```

If your `command` and/or `pattern` are the same throughout a test case or throughout a test suite, you can set the test case instance variables `@default_show_command` and/or `@default_output_pattern` which serve as defaults for these parameters:

```
@default_show_command = 'show run interface all | include "interface" | no-more'
assert_output_match(pattern: /interface port-channel 10/)
refute_output_match(pattern: /interface port-channel 11/)
refute_output_match(pattern: /interface port-channel 12/)
assert_output_match(pattern: /interface port-channel 13/)
```

### CODE REVIEW

**Comment by glennmatthews** [comment]
Example failure message when using these methods:

```
  1) Failure:
TestSnmpServer#test_snmpserver_packetsize_set_default [tests/test_snmpserver.rb:261]:
Expected /snmp-server packetsize (\d+)/ to NOT match output of '"show run snmp all | no-more"':
show run snmp all | no-more

!Command: show running-config snmp all
!Time: Fri Oct  9 15:23:50 2015

version 7.0(3)I2(1)
snmp-server aaa-user cache-timeout 3600
snmp-server packetsize 1500
snmp-server protocol enable 
snmp-server tcp-session auth
snmp-server globalEnforcePriv
snmp-server user admin network-admin auth md5 0x7a17ac9376fb63ba8321574872784681 priv 0x7a17ac9376fb63ba8321574872784681 localizedkey
snmp-server user devops network-admin auth md5 0x5268db65f8c4edd2c6c45fa3416d7ce8 priv 0x5268db65f8c4edd2c6c45fa3416d7ce8 localizedkey
snmp-server user dummy_user network-operator 
rmon event 1 log trap public description FATAL(1) owner PMON@FATAL
rmon event 2 log trap public description CRITICAL(2) owner PMON@CRITICAL
rmon event 3 log trap public description ERROR(3) owner PMON@ERROR
rmon event 4 log trap public description WARNING(4) owner PMON@WARNING
rmon event 5 log trap public description INFORMATION(5) owner PMON@INFO
snmp-server enable traps entity entity_mib_change
snmp-server enable traps entity entity_module_status_change
snmp-server enable traps entity entity_power_status_change
snmp-server enable traps entity entity_module_inserted
snmp-server enable traps entity entity_module_removed
snmp-server enable traps entity entity_unrecognised_module
snmp-server enable traps entity entity_fan_status_change
snmp-server enable traps entity entity_power_out_change
snmp-server enable traps link linkDown
snmp-server enable traps link linkUp
snmp-server enable traps link extended-linkDown
snmp-server enable traps link extended-linkUp
snmp-server enable traps link cieLinkDown
snmp-server enable traps link cieLinkUp
snmp-server enable traps link delayed-link-state-change
no snmp-server enable traps callhome event-notify
no snmp-server enable traps callhome smtp-send-fail
no snmp-server enable traps cfs state-change-notif
no snmp-server enable traps cfs merge-failure
snmp-server enable traps rf redundancy_framework
no snmp-server enable traps aaa server-state-change
snmp-server enable traps license notify-license-expiry
snmp-server enable traps license notify-no-license-for-feature
snmp-server enable traps license notify-licensefile-missing
snmp-server enable traps license notify-license-expiry-warning
snmp-server enable traps upgrade UpgradeOpNotifyOnCompletion
snmp-server enable traps upgrade UpgradeJobStatusNotify
no snmp-server enable traps feature-control FeatureOpStatusChange
no snmp-server enable traps sysmgr cseFailSwCoreNotifyExtended
snmp-server enable traps rmon risingAlarm
snmp-server enable traps rmon fallingAlarm
snmp-server enable traps rmon hcRisingAlarm
snmp-server enable traps rmon hcFallingAlarm
no snmp-server enable traps config ccmCLIRunningConfigChanged
no snmp-server enable traps snmp authentication
no snmp-server enable traps link cisco-xcvr-mon-status-chg
no snmp-server enable traps vtp notifs
no snmp-server enable traps vtp vlancreate
no snmp-server enable traps vtp vlandelete
no snmp-server enable traps bridge newroot
no snmp-server enable traps bridge topologychange
no snmp-server enable traps stpx inconsistency
no snmp-server enable traps stpx root-inconsistency
no snmp-server enable traps stpx loop-inconsistency
snmp-server enable traps entity entity_sensor
snmp-server enable traps entity cefcMIBEnableStatusNotification
no snmp-server enable traps system Clock-change-notification
snmp-server enable traps generic coldStart
snmp-server enable traps generic warmStart
no snmp-server enable traps feature-control ciscoFeatOpStatusChange
snmp-server enable traps storm-control cpscEventRev1


n9k-108.cisco.com#
```

File: docs/README-develop-node-utils-APIs.md [hunk 1]
```diff
@@ -179,7 +179,9 @@ end
 
 * A minitest should be created to validate the new APIs. Minitests are stored in the tests directory: `tests/`
 
-* Tests may use `config(cmd1, cmd2, ...)` and `@device.cmd("show ...")` to access the CLI directly set up tests and validate expected outcomes. The tests directory contains many examples of how these are used.
+* Tests may use `config(cmd1, cmd2, ...)` to access the CLI directly to set up tests.
+
+* Tests may use `@device.cmd("show ...")` or the helper methods `assert_show_match` and `refute_show_match` to validate expected outcomes in the CLI. The tests directory contains many examples of how these are used.
```
Comment (robert-w-gries) [inline docs/README-develop-node-utils-APIs.md:184]: Can you add a line to the readme describing @default_show_command and @default_output_pattern?

File: tests/basetest.rb [hunk 2]
```diff
@@ -100,7 +100,40 @@ def teardown
   end
 
   def config(*args)
-    @device.cmd("configure terminal\n" + args.join("\n") + "\nend")
+    # Send the entire config as one string but be sure not to return until
+    # we are safely back out of config mode, i.e. prompt is
+    # 'switch#' not 'switch(config)#' or 'switch(config-if)#' etc.
+    @device.cmd('String' => "configure terminal\n" + args.join("\n") + "\nend",
+                'Prompt' => /^[^()]+[$%#>] \z/n)
```
Comment (robert-w-gries) [inline tests/basetest.rb]: This line is a little cryptic to me. What happens when the regex is a match?

File: tests/test_interface.rb [hunk 3]
```diff
@@ -586,8 +582,7 @@ def negotiate_auto_helper(interface, default, cmd_ref)
                  "Error: #{inf_name} negotiate auto value not #{default}")
 
     pattern = cmd_ref.test_config_get_regex[default ? 1 : 0]
-    assert_interface_match_line(interface.name, pattern,
-                                "#{inf_name} no negotiate auto cfg mismatch")
+    assert_show_match(pattern: pattern)
```
Comment (robert-w-gries) [inline tests/test_interface.rb:585]: Shouldn't the command parameter be specified?

```
assert_show_match(command: cmd, pattern: pattern)
```

**Comment by robert-w-gries** [comment]
Can you post code coverage results of these changes?  Does it reduce the code coverage at all?

File: docs/README-develop-node-utils-APIs.md (hunk 1 above)
Comment (glennmatthews) [inline docs/README-develop-node-utils-APIs.md:184]: I'm increasingly thinking what we really need is a separate README specifically for details about writing good quality Minitest scripts. Would you mind if I defer this until then?

File: tests/basetest.rb (hunk 2 above)
Comment (glennmatthews) [inline tests/basetest.rb]: `cmd()` doesn't return until it sees a string that matches the given Prompt regex. This regex matches a line that has a series of any characters that do NOT include '(' or ')', ending in one of the standard prompt characters `$%#>` followed by a single space.

The exclusion of '(' and ')' means that we will not match config-mode prompt lines like `switch(config)#` or `n9k-108(config-if)#` - we will only match (and be correctly 'done' with the config command) when we get an exec-mode prompt like `switch#` or `uut.cisco.com#`.

I was hoping the comment above made this clear enough  but I guess not. What can I add to make this more self-evident?

File: tests/test_interface.rb (hunk 3 above)
Comment (glennmatthews) [inline tests/test_interface.rb:585]: This is where the magic comes in with `@default_show_command` - if you look below, both `test_negotiate_auto_portchannel` and `test_negotiate_auto_ethernet` are setting `@default_show_command`, so if we don't pass a `command` parameter to this method, that default show command is what gets used.

Is this being too clever for my own good?

**Comment by glennmatthews** [comment]
Shouldn't have any impact on code coverage but just to be sure:

Baseline: `1836 / 1927 LOC (95.28%) covered.`
My branch: `1826 / 1921 LOC (95.05%) covered.`

We have a couple of intermittent failures at present so these aren't exactly identical runs but close enough.

**Comment by glennmatthews** [comment]
Crud, still seeing the test_interface_ipv4_all_interfaces failure:

```
  1) Failure:
TestInterface#test_interface_ipv4_all_interfaces [/home/admin/Desktop/cisco-network-node-utils/tests/test_interface.rb:73]:
Found no matching lines in:
no switchport
n9k-108.cisco.com(config-if)# ip address 10.7.1.1/15
n9k-108.cisco.com(config-if)# description Company B
n9k-108.cisco.com(config-if)# end
n9k-108.cisco.com# configure terminal
Enter configuration commands, one per line. End with CNTL/Z.
n9k-108.cisco.com(config)# interface loopback0
n9k-108.cisco.com(config-if)# ip address 11.7.1.1/15
n9k-108.cisco.com(config-if)# description 233KLDK
n9k-108.cisco.com(config-if)# end
n9k-108.cisco.com# .
Expected [] to not be empty.
```

**Comment by glennmatthews** [comment]
Fixed config() API to properly fix the interface failure - Telnet.new() takes 'Prompt' but Telnet.cmd() takes 'Match' - and silently ignores 'Prompt' as a parameter. Doh.

**Comment by glennmatthews** [comment]
1830 / 1921 LOC (95.26%) covered.

File: docs/template-test_router.rb [hunk 4]
```diff
@@ -60,23 +59,23 @@ def test_router_create_destroy_multiple
...
+
     s = @device.cmd("show runn | i 'router X__RESOURCE_NAME__X'")
     assert_match(s, /^router X__RESOURCE_NAME__X #{id1}$/)
     assert_match(s, /^router X__RESOURCE_NAME__X #{id2}$/)
 
     rtr1.destroy
-    s = @device.cmd("show runn | i 'router X__RESOURCE_NAME__X #{id1}'")
-    refute_match(s, /^router X__RESOURCE_NAME__X #{id1}$/,
-                 "Error: failed to destroy router X__RESOURCE_NAME__X #{id1}")
+    refute_show_match(pattern: /^router X__RESOURCE_NAME__X #{id1}$/,
```
Comment (mikewiebe) [inline docs/template-test_router.rb:69]: Glad you fixed this :)  One of the hackathon guys ran into this a few mins ago.

File: docs/README-develop-node-utils-APIs.md (hunk 1 above)
Comment (robert-w-gries) [inline docs/README-develop-node-utils-APIs.md:184]: :+1:

File: tests/test_interface.rb (hunk 3 above)
Comment (robert-w-gries) [inline tests/test_interface.rb:585]: Nope not a problem :)

**Comment by robert-w-gries** [comment]
:+1:

**Comment by mikewiebe** [comment]
:+1:

**Pull request merged**